File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.go
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func earliestFinishTime(landStartTime []int, landDuration []int, waterStartTime []int, waterDuration []int) int {
	const INF = int(1e18)

	ans, minLand, minWater := INF, INF, INF

	for i := 0; i < len(landStartTime); i++ {
		minLand = min(minLand, landStartTime[i]+landDuration[i])
	}

	for i := 0; i < len(waterStartTime); i++ {
		ans = min(ans, max(minLand, waterStartTime[i])+waterDuration[i])
	}

	for i := 0; i < len(waterStartTime); i++ {
		minWater = min(minWater, waterStartTime[i]+waterDuration[i])
	}

	for i := 0; i < len(landStartTime); i++ {
		ans = min(ans, max(minWater, landStartTime[i])+landDuration[i])
	}

	return ans
}

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.py
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        ans = mln = mnw = float("inf")

        for i in range(len(landStartTime)):
            mln = min(mln, landStartTime[i] + landDuration[i])

        for i in range(len(waterStartTime)):
            ans = min(ans, max(mln, waterStartTime[i]) + waterDuration[i])

        for i in range(len(waterStartTime)):
            mnw = min(mnw, waterStartTime[i] + waterDuration[i])

        for i in range(len(landStartTime)):
            ans = min(ans, max(mnw, landStartTime[i]) + landDuration[i])

        return ans
        